### ISSUE TITLE
Scylla exits with exception on schema migration of dense table with regular columns

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#2583]
*Installation details*
Scylla version (or git commit hash): 2.0+

```
ERROR 2017-07-12 11:20:54,669 [shard 0] seastar - Exiting on unhandled exception: std::invalid_argument (unknown type: org.apache.cassandra.db.marshal.regular
)
```

```
(gdb) bt
#0  0x00007ffff7246e9d in __cxa_throw () from /lib64/libstdc++.so.6
#1  0x0000000000cc2a79 in abstract_type::parse_type (name=...) at types.cc:3096
#2  0x0000000000fd5938 in db::marshal::type_parser::get_abstract_type (compare_with=...) at db/marshal/type_parser.cc:196
#3  0x0000000000fd6f2e in db::marshal::type_parser::do_parse (this=this@entry=0x7fffffffc9d0, multicell=multicell@entry=true) at db/marshal/type_parser.cc:106
#4  0x0000000000fd73a8 in db::marshal::type_parser::parse (this=0x7fffffffc9d0) at db/marshal/type_parser.cc:71
#5  db::marshal::type_parser::parse (str=...) at db/marshal/type_parser.cc:63
#6  0x0000000000e84639 in db::schema_tables::parse_type (str=...) at db/schema_tables.cc:2504
#7  0x0000000000f0a794 in db::legacy_schema_migrator::migrator::read_table(db::legacy_schema_migrator::migrator::keyspace&, seastar::basic_sstring<char, unsigned int, 15u>, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> >&&)#1}::operator()(std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> >&&) const::{lambda()#1}::operator()() const (__closure=__closure@entry=0x7fffffffcd50) at db/legacy_schema_migrator.cc:264
#8  0x0000000000f27ed8 in db::legacy_schema_migrator::migrator::read_table(db::legacy_schema_migrator::migrator::keyspace&, seastar::basic_sstring<char, unsigned int, 15u>, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> >&&)#1}::operator()(std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> >&&) const (__closure=__closure@entry=0x600005595850, t=...) at db/legacy_schema_migrator.cc:277
#9  0x0000000000f2e3ed in seastar::apply_helper<db::legacy_schema_migrator::migrator::read_table(db::legacy_schema_migrator::migrator::keyspace&, seastar::basic_sstring<char, unsigned int, 15u>, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> >&&)#1}, std::tuple<std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> > >&&, std::integer_sequence<unsigned long, 0ul> >::apply({lambda(std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> >&&)#1}&&, std::tuple<std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> > >) (args=..., func=...)
    at ./seastar/core/apply.hh:36
#10 seastar::apply<db::legacy_schema_migrator::migrator::read_table(db::legacy_schema_migrator::migrator::keyspace&, seastar::basic_sstring<char, unsigned int, 15u>, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> >&&)#1}, std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> > >(db::legacy_schema_migrator::migrator::read_table(db::legacy_schema_migrator::migrator::keyspace&, seastar::basic_sstring<char, unsigned int, 15u>, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> >&&)#1}&&, std::tuple<std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> > >&&) (args=..., func=...) at ./seastar/core/apply.hh:44
#11 seastar::do_void_futurize_apply_tuple<db::legacy_schema_migrator::migrator::read_table(db::legacy_schema_migrator::migrator::keyspace&, seastar::basic_sstring<char, unsigned int, 15u>, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> >&&)#1}, std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<seastar::shared_ptr<cql3::untyped_result_set> >, seastar::future<db::schema_tables::legacy::schema_mutations> > >(std::result_of&&, std::tuple<(db::legacy_schema_migrator::migrator::read_table(db::legacy_schema_migrator::migrator::keyspace&, seastar::basic_sstring<char, unsigned int, 15u>, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(std::tuple<seastar::future<seastar::shared_ptr<cql3::untyped---Type <return> to continue, or q <return> to quit---Quit
(gdb) select-frame 5
(gdb) p str
$1 = (const seastar::sstring &) @0x7fffffffca30: {u = {external = {str = 0x72616c75676572 <error: Cannot access memory at address 0x72616c75676572>, 
      size = 1701535598, pad = 121 'y'}, internal = {str = "regular\000n_key\000x", size = 7 '\a'}}, static npos = <optimized out>}
(gdb) 

```

The problem is that the code tries to interpret `type` column as a type, while it should look at `validator`.

**Comment by elcallio** [comment]
:-( Cut and paste error.


Den 2017-07-12 kl. 11:28, skrev Tomasz Grabiec:

**Comment by tgrabiec** [comment]
Fixed by #2587
